### PR TITLE
Week06

### PR DESCRIPTION
**Pull request by wolves**
Summary
---------------------------------------------------
*Summarize what is included in this PR?*

This pull request contains a set of tests with 94.4% coverage for each of the individual program components provided including but not limited to errors, validation, and concurrency.

Code Description
------------------------------------------
*Briefly describe what your code intends to do*

### Completed_Product

This provides a basic structure around a `Product` associating it with the `Employee` that created it. As such, the test for it is relatively straightforward. There is a test table that exercises the valid path, the invalid employee condition, and the invalid product quantity condition.

```Go
validProduct := stubs{Product: p1, Employee: e}
noEmpProd := stubs{Product: p1}
noQtyProd := stubs{Product: p0, Employee: e}

[...]

{name: "a valid product is IsValid", stubs: validProduct, expErr: false},
{name: "no employee product IsInvalid", stubs: noEmpProd, expErr: true},
{name: "no quantity product IsInvalid", stubs: noQtyProd, expErr: true},
```

### Employee 

An `Employee` represents the worker that will be passed into the `Product.Build` method when there are `Jobs()` present via the `Manager`. Its method `IsValid()` verifies that the employee's integer representation is `> 0` and its private `work()` function is used by the `Manager` to start the directed number of employees so that they are prepared to receive incoming jobs.

The tests for this functionality were implemented as a table test to verify the employee validation with positive, negative, and zero values.

```Go
{name: "with valid employee", e: Employee(2), expErr: false},
{name: "with zero employee", e: Employee(0), expErr: true},
{name: "with negative employee", e: Employee(-2), expErr: true},
```

As part of the work function, the `Employee` also communicates any errors to the `Manager.Errors()` channel. This was tested by starting a worker, passing a `Product` with an invalid quantity to the `Manager.Jobs()` channel, and then verifying the resulting error was the correct type.

### Product

In our program, the `Product` contains a `Quantity` and a private `builtBy` field that holds the integer ID of the associated `Employee` that it was built by. Its exposed methods are `Build` which "produces" the product and associates the `Employee` directed to build it, `IsValid` which validates that a `Quantity` exists on the calling `Product`, and `IsBuilt` which verifies the `Product` is valid as well as the associated `Employee`.

The functionality of the exposed methods was verified using a set of table tests that represented their potential branching outcomes.

```Go
// BuiltBy
{name: "with an existing employee", stubs: stubs{p, e1}, want: e1},
{name: "with no employee", stubs: stubs{p, e0}, want: e0},

// Build
{name: "with valid inputs", stubs: stubs{Quantity: 1}, args: args{e: e1}, wantErr: false},
{name: "with no product", stubs: stubs{Quantity: 0}, args: args{e: e1}, wantErr: true},
{name: "with no employee", stubs: stubs{Quantity: 1}, args: args{e: e0}, wantErr: true},

// IsValid
{name: "with valid inputs", fields: fields{Quantity: 6}, wantErr: false},
{name: "with invalid inputs", fields: fields{Quantity: 0}, wantErr: true},

// IsBuilt
validInputs := stubs{Quantity: 1, Employee: Employee(1)}
noQtyInputs := stubs{Quantity: 0, Employee: Employee(1)}
noBuilderStubs := stubs{Quantity: 1, Employee: Employee(0)}

[...]

{name: "with valid inputs", stubs: validInputs, wantErr: false},
{name: "with no quantity", stubs: noQtyInputs, wantErr: true},
{name: "with no builder", stubs: noBuilderStubs, wantErr: true},
```

### Errors

The provided custom errors are as follows and were tested to verify their corresponding `Error()` return values.

```Go
type ErrInvalidQuantity int

type ErrProductNotBuilt string

type ErrInvalidEmployee int

type ErrInvalidEmployeeCount int

type ErrManagerStopped struct{}
```

### Manager

The `Manager`'s primary function is to integrate the rest of the pieces together resulting in the coordinated assembly of requested products via the employee worker pool which it spins up according to the provided employee count passed into the `Start` method. Valid products are then passed into the `Job()` channel via `Manager.Assign`. The manager also contains channels for `errs`, `jobs`, `completed` products, a `quit` trigger which sets its `stopped` state field value.

`Manager` was tested with a table test composed of valid components, invalid `Product` inputs, and an invalid `Employee` count. Each set was checked and verified for both the resulting `error` as well as `CompletedProducts`.

```Go
{
  name:        "with valid products and workers",
  workerCount: 2,
  prodQtys:    validProdList,
  expCpCount:  len(validProdList),
  expErr:      nil,
}, {
  name:        "with an invalid product list",
  workerCount: 1,
  prodQtys:    noQtyProdList,
  expCpCount:  0,
  expErr:      ErrInvalidQuantity(0),
}, {
  name:        "with an invalid emp count",
  workerCount: 0,
  prodQtys:    validProdList,
  expCpCount:  0,
  expErr:      ErrInvalidEmployeeCount(0),
},
```

Essay/Commentary
------------------------------------------
*A long-form writeup covering: Difficulties encountered / Noteworthy choices made / Other things that stood out to you*

This week's assignment was challenging at moments because of the elusiveness of timely Goroutine output and my associated inexperience, That being said, having pushed through and gained the resulting knowledge I have, I feel as though the hill was a much less strenuous one to climb. It is a promising feeling especially since toward the middle 90% coverage area of these tests I was beginning to feel as though I was understanding and able to correctly manipulate the Goroutines and corresponding channels for the sake of my tests.

I truly wanted to achieve 100% coverage simply as a personal challenge and with a bit more time I believe I could have, nonetheless, I was at least able to reach the mid 90's. The time spent manipulating different ways of testing certain elements, only to have to check and rework them when they behaved differently than I anticipated may have cost me the short-term goal I had imposed on myself, but in its place, I was provided a much more valuable gain; A deeper understanding of Go's concurrency than I had prior.

The familiarity gained relating to Goroutines is also enough to make me quite aware that continued, diligent exercise is still required on my part if I am to find true comfort when implementing and/or working with them though. I look forward to the coming weeks to do just that.

### CODE REVIEW

**Comment by wolves** [comment]
Unfortunately, I didn't have the time last night to go back and update some of the tests with the updated constraints regarding accessing private functions for unit testing. This will be most evident in regards to the `Employee.work()` function

File: week06/errors_test.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
+package week06
+
+import (
+	"fmt"
+	"testing"
+)
+
+func Test_Errors(t *testing.T) {
```
Comment (markbates) [inline week06/errors_test.go:8]: This feels like it could be wrapped into a table driven test.

File: week06/manager_test.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+		ps = append(ps, Product{Quantity: q})
+	}
+	return ps
+}
+
+// Spin up the employee pool based on the provided worker count
+func startUpWorkerPool(t testing.TB, m *Manager, n int) error {
+	t.Helper()
+
+	err := m.Start(n)
```
Comment (markbates) [inline week06/manager_test.go:93]: You need to make sure these resources are cleaned up. Using the `Cleanup` method on `testing.T` you can ensure that regardless of what happens in the test itself, the resource will be cleaned by the go test runner.

```suggestion
	err := m.Start(n)
	t.Cleanup(m.Stop)
```

**Review by markbates — COMMENTED** [review]
Excellent work!

**Pull request merged**